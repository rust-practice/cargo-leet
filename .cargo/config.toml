[alias]
i = "install --path . --features=tool"
lg = "run --features=tool -- leet gen"
