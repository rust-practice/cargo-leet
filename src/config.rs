pub struct Config {}

impl Config {
    pub const LEETCODE_URL: &str = "https://leetcode.com/problems/";
    pub const LEETCODE_GRAPH_QL: &str = "https://leetcode.com/graphql/";
}
