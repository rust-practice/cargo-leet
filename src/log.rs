use anyhow::Context;
use log::LevelFilter;
use log4rs::{
    append::console::ConsoleAppender,
    config::{Appender, Root},
    encode::pattern::PatternEncoder,
};

pub fn init_logging(log_level: LevelFilter) -> anyhow::Result<()> {
    let stdout = ConsoleAppender::builder()
        .encoder(Box::new(PatternEncoder::new("{h({M} {l} - {m})}{n}")))
        .build();

    let config = log4rs::Config::builder()
        .appender(Appender::builder().build("stdout", Box::new(stdout)))
        .build(Root::builder().appender("stdout").build(log_level))?;
    log4rs::init_config(config).context("Failed to initialize logging")?;
    Ok(())
}
