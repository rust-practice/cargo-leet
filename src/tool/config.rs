pub(crate) struct Config {}

impl Config {
    // assumed in the code using them URLs Must end with trailing "/"
    pub(crate) const LEETCODE_PROBLEM_URL: &str = "https://leetcode.com/problems/";
    pub(crate) const LEETCODE_GRAPH_QL: &str = "https://leetcode.com/graphql/";
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn problem_url_ends_with_slash() {
        assert!(Config::LEETCODE_PROBLEM_URL.ends_with('/'));
    }

    #[test]
    fn graph_ql_url_ends_with_slash() {
        assert!(Config::LEETCODE_GRAPH_QL.ends_with('/'));
    }
}
