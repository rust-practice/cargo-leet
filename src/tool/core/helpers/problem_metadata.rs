use crate::tool::config::Config;
use anyhow::Context;
use log::info;

use super::{get_response, local_store::path_local_store_problem_metadata};

pub(crate) mod data_structure;

pub(crate) fn get_problem_metadata(
    title_slug: &str,
) -> anyhow::Result<data_structure::ProblemMetadata> {
    info!("Attempting to get problem metadata");
    get_problem_metadata_response(title_slug)?.into_problem_metadata()
}

fn get_problem_metadata_response(
    title_slug: &str,
) -> anyhow::Result<data_structure::ProblemMetaDataResponse> {
    get_response(
        title_slug,
        local_store_request_problem_metadata,
        external_request_problem_metadata,
    )
}

fn local_store_request_problem_metadata(title_slug: &str) -> anyhow::Result<String> {
    let path = path_local_store_problem_metadata(title_slug);
    std::fs::read_to_string(&path).with_context(|| format!("failed to read string from {path:?}"))
}

fn external_request_problem_metadata(title_slug: &str) -> anyhow::Result<String> {
    info!("[External] Going to send request for problem meta data for problem with title: {title_slug}");
    ureq::get(Config::LEETCODE_GRAPH_QL)
        .send_json(ureq::json!({
            "query": r"query consolePanelConfig($titleSlug: String!) {
            question(titleSlug: $titleSlug) {
                questionFrontendId
                questionTitle
                exampleTestcaseList
            }
        }",
            "variables":{"titleSlug": title_slug},
            "operationName":"consolePanelConfig"
        }))
        .context("failed to get request for code_snippet failed")?
        .into_string()
        .context("failed to convert response into String")
}

#[cfg(test)]
mod tests;
