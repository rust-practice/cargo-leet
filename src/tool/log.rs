#![cfg(feature = "tool")]
use env_logger::Builder;
use log::LevelFilter;

pub fn init_logging(level: LevelFilter) -> anyhow::Result<()> {
    Builder::new().filter(None, level).try_init()?;
    Ok(())
}
