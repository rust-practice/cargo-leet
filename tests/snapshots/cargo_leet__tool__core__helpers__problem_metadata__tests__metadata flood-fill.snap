---
source: src/tool/core/helpers/problem_metadata.rs
expression: get_problem_metadata(title_slug).unwrap()
---
ProblemMetadata {
    id: 733,
    title: "Flood Fill",
    example_test_case_list: [
        "[[1,1,1],[1,1,0],[1,0,1]]\n1\n1\n2",
        "[[0,0,0],[0,0,0]]\n0\n0\n0",
    ],
}
